### PR TITLE
Refactor AST traversing

### LINKED ISSUE #541: `traverse` function does not perform a complete traverse of AST
For example, traversing `ASTStatementLet` currently looks this way:
https://github.com/tact-lang/tact/blob/cfcb4f04ce5155e0d6638c0b6b1f7a920f5c1792/src/grammar/ast.ts#L759-L761
while it should look like this:
```ts
if (node.kind === "statement_let") {
    traverse(node.name, callback);
    if (node.type) traverse(node.type, callback);
    traverse(node.expression, callback);
}
```
becase `AstStatementLet` includes two other AST nodes apart from `expression`:
https://github.com/tact-lang/tact/blob/cfcb4f04ce5155e0d6638c0b6b1f7a920f5c1792/src/grammar/ast.ts#L188-L195

And this problem exists for almost all ASTNode types in `traverse` function, which can cause problems in `resolveDescriptors`, especially if we want to perform some new traversals as for resolving #533.

### PR DESCRIPTION
**Pull request by Gusarich**
Closes #541 

- [ ] ~~I have updated CHANGELOG.md~~
- [ ] ~~I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER~~
- [ ] ~~I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases~~
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

File: src/grammar/ast.ts [hunk 1]
```diff
@@ -217,14 +217,14 @@ export type AstStatementAssign = {
...
 export type AstAugmentedAssignOperation =
-    | "+"
-    | "-"
-    | "*"
-    | "/"
-    | "%"
-    | "|"
-    | "&"
-    | "^";
+    | { kind: "augmentedassign_operation"; operation: "+" }
```
Comment (anton-trunov) [inline src/grammar/ast.ts]: I'm actually not at all convinced we should add these tags to operators and attributes here and below. For instance, I need to collect attributes for a function, I'd traverse only the function definitions and native function declarations. And if you need some info on augmented assignment operator you 100% need to look at the AST node that actually includes meaningful information like the path and the right-hand side expression. I propose to revert this part of the refact and only introduce things like this if there is ever a use case

File: src/grammar/ast.ts [hunk 2]
```diff
@@ -626,23 +632,29 @@ export type AstReceiverKind =
...
-    | AstConstantDecl;
+    | AstModuleItem
+    | AstTypeDecl
+    | AstContractDeclaration
+    | AstTraitDeclaration
+    | AstStatement
+    | AstAugmentedAssignOperation
+    | AstType
+    | AstTypeId
+    | AstOptionalType
```
Comment (anton-trunov) [inline src/grammar/ast.ts]:
```
    | AstTypeId
    | AstOptionalType
    | AstMapType
    | AstBouncedMessageType
```

should be covered by `AstType`

**Comment by anton-trunov** [comment]
What is the minimal change that is needed to resolve #533?

**Comment by Gusarich** [comment]
> What is the minimal change that is needed to resolve #533?

I think the change of `traverse` function is enough and we don't need other refactors for that.
Should I just revert all the refactors that I did here and only leave `traverse` function rewrite?

**Comment by anton-trunov** [comment]
yeah, let's keep the minimal changes

**Comment by anton-trunov** [comment]
btw, I broke already broken CI for Windows, so please ignore that for now

**Review by anton-trunov — COMMENTED** [review]
LGTM. Let's also remove `iterators.ts` from the list of files ignored by Knip: https://github.com/tact-lang/tact/blob/eb9ec73920be7830a072b98c87f1e79eb23949ef/knip.json#L7, since it's now being used in other modules within Tact

**Review by anton-trunov — DISMISSED** [review]

**Comment by anton-trunov** [comment]
the unused deps marked by Knip need to be made `@public`, it should fix the failing check:

```
$ knip
Unused exports (4)
foldExpressions    function  src/grammar/iterators.ts:182:17
forEachStatement   function  src/grammar/iterators.ts:393:17
foldStatements     function  src/grammar/iterators.ts:501:17
forEachExpression  function  src/grammar/iterators.ts:8:17
```

**Pull request merged**